### PR TITLE
Implementing POM

### PR DESCRIPTION
**Pull request by KurisuLim**
### JAVA AUTOMATION A4  Homework-Refactor your tests using Page Object Model
1. Refactor your tests using POM.
2. Create Base Class which will include all generic methods that you gonna use in each page object class
3. Create Classes for each page(Login, Sign Up, PDP, PLP, Search)
In your tests call those POM methods

### CODE REVIEW

File: src/test/java/pages/BasePage.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package pages;
+
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.openqa.selenium.support.ui.WebDriverWait;
+
+public class BasePage {
+    public WebDriver driver;
+    public WebDriverWait wait;
```
Comment (mobilman89) [inline src/test/java/pages/BasePage.java:10]: make a private

File: src/test/java/pages/BasePage.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package pages;
+
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.openqa.selenium.support.ui.WebDriverWait;
+
+public class BasePage {
+    public WebDriver driver;
```
Comment (mobilman89) [inline src/test/java/pages/BasePage.java:9]: make a protected

File: src/test/java/pages/HomePage.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+
+public class HomePage extends BasePage{
+
+    public HomePage(WebDriver driver) {
+        super(driver);
+    }
+
+    protected static String BASE_URL = "https://testkwidos.tk/" ;
```
Comment (mobilman89) [inline src/test/java/pages/HomePage.java:12]: I think this variable should be private final static

File: src/test/java/pages/HomePage.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+    //Solution for Homework 1
+    By loginButton = By.cssSelector("ul.user.nav");
```
Comment (mobilman89) [inline src/test/java/pages/HomePage.java:15]: make a private

File: src/test/java/pages/LoginPage.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        //Will enable on the next homework
+        //AjaxElementLocatorFactory PageFactory = new AjaxElementLocatorFactory(driver, 10);
+        //org.openqa.selenium.support.PageFactory.initElements(driver, this);
+    }
+
+    protected By email = By.name("email");
+    protected By password = By.name("password");
+    protected By loginBtn = By.tagName("button");
+    protected By titleText = By.tagName("h5");
+    protected By errorText = By.className("help-block");
```
Comment (mobilman89) [inline src/test/java/pages/LoginPage.java:19]: all locators should be private static final

File: src/test/java/tests/LoginTest.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        String actual = driver.getTitle();
+        Assert.assertEquals(actual, expected);
+    }
+
+
+    @Test
+    public void emptyLogin(){
+        page.loginToTestkwidos("","");
+        Assert.assertTrue(page.isErrorVisible());
+        System.out.println("Empty Login");
```
Comment (mobilman89) [inline src/test/java/tests/LoginTest.java:32]: remove System.out.println

File: src/test/java/pages/Users.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
+package pages;
+
+public class Users {
```
Comment (mobilman89) [inline src/test/java/pages/Users.java:3]: better call this class 'User'

File: src/test/java/pages/Users.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    protected String firstName,
```
Comment (mobilman89) [inline src/test/java/pages/Users.java:4]: All variables should be a private and access to variables should achieve using getters and setters

File: src/test/java/pages/Users.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+            lastName,
+            businessName,
+            phone,
+            email,
+            password;
+
+    public Users emptySignUp() {
```
Comment (mobilman89) [inline src/test/java/pages/Users.java:11]: this method should placing in class UserFactory and each time return new User instance

File: src/test/java/pages/Users.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+        firstName = "";
+        lastName = "";
+        businessName = "";
+        phone = "";
+        email = "";
+        password = "";
+        return this;
+    }
+
+    public Users wrongSignUp() {
```
Comment (mobilman89) [inline src/test/java/pages/Users.java:21]: see comment above about UserFactory

File: src/test/java/pages/SignupPage.java [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+    public SignupPage(WebDriver driver) {
+        super(driver);
+        //Will enable on the next homework
+        //AjaxElementLocatorFactory PageFactory = new AjaxElementLocatorFactory(driver, 10);
+        //org.openqa.selenium.support.PageFactory.initElements(driver, this);
+    }
+
+    Users user = new Users();;
+    Faker faker = new Faker();
```
Comment (mobilman89) [inline src/test/java/pages/SignupPage.java:17]: I think that User and Faker are the same entities

File: src/test/java/pages/SignupPage.java [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+    protected By firstName = By.cssSelector("input[formcontrolname='firstName']");
+    protected By lastName = By.cssSelector("input[formcontrolname='lastName']");
+    protected By businessName = By.cssSelector("input[formcontrolname='businessName']");
+    protected By phone = By.cssSelector("input[formcontrolname='phone']");
+    protected By email = By.cssSelector("input[formcontrolname='email']");
+    protected By password = By.cssSelector("input[formcontrolname='password']");
```
Comment (mobilman89) [inline src/test/java/pages/SignupPage.java:24]: all locators should be private final

File: src/test/java/pages/SignupPage.java [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+    protected By termsOfService = By.cssSelector("div.card-content div:nth-child(8) div p-checkbox");
+    protected By createAccountBtn = By.tagName("button");
+    public By alert = By.cssSelector(".alert");
+
+    public void setFirstName(String fName){
```
Comment (mobilman89) [inline src/test/java/pages/SignupPage.java:29]: make a private

File: src/test/java/pages/SignupPage.java [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+        driver.findElement(firstName).sendKeys(fName);
+    }
+
+    public void setLastName(String lName){
```
Comment (mobilman89) [inline src/test/java/pages/SignupPage.java:33]: all methods which you use only in this class should be private

File: src/test/java/pages/SignupPage.java [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    public void checkboxTOS(){
+        driver.findElement(termsOfService).click();
+    }
+    public void clickCreateAccount(){
+        waitElementVisibilityAndClick(createAccountBtn);
+    }
+
+    public SignupPage emptySignup(){
```
Comment (mobilman89) [inline src/test/java/pages/SignupPage.java:60]: please combine methods emptySignup, wrongSignUp, wrongSignUp and validSignUp to one method signUp which accepts 'user' parameter

File: src/test/java/tests/LoginTest.java [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+    @BeforeMethod
+    public void loginPage(){
+        base = new BasePage(driver);
+        home = new HomePage(driver);
+        page = new LoginPage(driver);
+        home.openHomePage()
+                .toLoginPage();
+        String expected = "Kwidos";
+        String actual = driver.getTitle();
+        Assert.assertEquals(actual, expected);
```
Comment (mobilman89) [inline src/test/java/tests/LoginTest.java:24]: you can change 3 lines to one. Assert.assertEquals(driver.getTitle(), "Kwidos");

File: src/test/java/tests/LoginTest.java [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+import pages.BasePage;
+import pages.HomePage;
+import pages.LoginPage;
+
+public class LoginTest extends BaseTest {
+    LoginPage page;
+    HomePage home;
+    BasePage base;
```
Comment (mobilman89) [inline src/test/java/tests/LoginTest.java:13]: make a private

File: src/test/java/tests/LoginTest.java [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+
+    @BeforeMethod
+    public void loginPage(){
+        base = new BasePage(driver);
```
Comment (mobilman89) [inline src/test/java/tests/LoginTest.java:17]: BasePage instance is not necessary

File: src/test/java/tests/SignUpTest.java [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+import pages.BasePage;
+import pages.HomePage;
+import pages.SignupPage;
+
+public class SignUpTest extends BaseTest {
+    SignupPage page;
+    HomePage home;
+    BasePage base;
```
Comment (mobilman89) [inline src/test/java/tests/SignUpTest.java:13]: private

Comment (mobilman89) [inline src/test/java/tests/SignUpTest.java:13]: remove BasePage instance

File: src/test/java/tests/SignUpTest.java [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+        home.openHomePage()
+                .toSignUpPage();
+        String expected = "Kwidos";
+        String actual = driver.getTitle();
+        Assert.assertEquals(actual, expected);
+    }
+
+    @Test
+    public void emptySignUp(){
+        page.emptySignup();
```
Comment (mobilman89) [inline src/test/java/tests/SignUpTest.java:29]: page.signup(UserFactory.getEmptyUser())

File: src/test/java/tests/SignUpTest.java [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+        System.out.println("Empty Signup Success");
+        //the Create Account Button is disabled if
+        //the credential is incorrect
+    }
+
+    @Test
+    public void invalidSignUp(){
+        page.wrongSignUp();
```
Comment (mobilman89) [inline src/test/java/tests/SignUpTest.java:37]: page.signup(UserFactory.getWrongUser())

File: src/test/java/tests/SignUpTest.java [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+    @BeforeMethod
+    public void signUpPage(){
+        base = new BasePage(driver);
+        home = new HomePage(driver);
+        page = new SignupPage(driver);
+        home.openHomePage()
+                .toSignUpPage();
+        String expected = "Kwidos";
+        String actual = driver.getTitle();
+        Assert.assertEquals(actual, expected);
```
Comment (mobilman89) [inline src/test/java/tests/SignUpTest.java:24]: remove assert from BeforeMethod! it is not test

File: src/test/java/tests/SignUpTest.java [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+        page.emptySignup();
+        System.out.println("Empty Signup Success");
+        //the Create Account Button is disabled if
+        //the credential is incorrect
+    }
+
+    @Test
+    public void invalidSignUp(){
+        page.wrongSignUp();
+        System.out.println("Invalid Signup Success");
```
Comment (mobilman89) [inline src/test/java/tests/SignUpTest.java:38]: Where is assert in test?

File: src/test/java/tests/SignUpTest.java [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+        String actual = driver.getTitle();
+        Assert.assertEquals(actual, expected);
+    }
+
+    @Test
+    public void emptySignUp(){
+        page.emptySignup();
+        System.out.println("Empty Signup Success");
+        //the Create Account Button is disabled if
+        //the credential is incorrect
```
Comment (mobilman89) [inline src/test/java/tests/SignUpTest.java:32]: What does verify your test?

Comment (KurisuLim) [reply src/test/java/tests/SignUpTest.java:32]: @mobilman89
It's impossible to write an assertion if the signup page does not let you do a negative test case.

Comment (mobilman89) [reply src/test/java/tests/SignUpTest.java:32]: Why not? you can check that you will not sign up or you get an error message

**Pull request merged**